### DISCUSSION TITLE
Getting odometry data from Optical Flow sensor for SLAM and NAV2

### DISCUSSION THREAD

**Discussion by Rak-r** [iRobotEducation/create3_docs#493] [Questions] [answered]
### How are you connecting to your Create 3?

Wi-Fi (ROS 2)

### Which version of ROS 2 is installed on your computer?

None

### Which firmware version is installed on your robot?

None

### Which RMW is your robot running?

CycloneDDS

### Are there multiple Create 3 robots connected to your network?

None

### Is multicast enabled?

None

### What is the Adapter Board's USB/BLE Toggle currently switched to?

None

### Describe your question.

I am having an optical flow sensor which is similar to the mouse message of iRobot. I have managed to setup the connection using serial and getting the data in x,y pixel co-ordinates. Converted the incoming data to actual distance X and Yin meters. Z corodinate is fixed (sensor mount height). Now I added the X,Y and Z distance in odometry translation message field. But I feel there might be something wrong. Need some guidance how can take the data and setup the slam and Nav2 stack for my robot.?

I saw the iRobot has a mouse message which seems some similar approach. It would be great if anyone could provide some insight.

I know the iRobot community is well supported and might be deals with iRobot related queries. However,It would be great if anyone could provide some insight and guidance.

**Comment by alsora** [comment, marked as answer]
Hi @Rak-r,
a common error when implementing dead-reckoning using a mouse sensor consists of not taking into account that the sensor location may not coincide with the robot center of mass.
For example, when the robot turns in place, its x/y/z location shouldn't change, but the mouse will detect movements.

To take that into account, you need to make sure to express the differential motion in the correct reference frame.
You can find some papers online detailing the necessary equations, such as:
```
Sekimori, D. and Miyazaki, F., 2007. Precise dead-reckoning for mobile robots using multiple optical mouse sensors. In Informatics in Control, Automation and Robotics II (pp. 145-151). Springer Netherlands.
```

**Reply by Rak-r** [reply]
Thanks for the quick reply. My robot is Ackermann-steered, so no in place turns. What I have done is taking the x,y pixel data as the sensor moves and converting to Actual distances by accumulating every last received data to get the right value and added to odom message field as explained above. Now I am bit clueless, how can I apply EKF to it for filtered odom with covariance because I have only translation fields and no orientation fields. 

If I can get orientation fields, how? is that even necessary or I can fill those with covariances. I am bit new to this and till now I managed to reach till above. Is there any example/reference of anybody using mouse or any optical flow sensor based robot?

I will definitely go through the paper you recommended.

**Reply by alsora** [reply]
> My robot is Ackermann-steered, so no in place turns.

It doesn't really matter which kinematic model your robot has, I simply made an example to show that when the mouse is not located at the center of mass of the robot (or wherever your robot local reference frame is located), then the motion detected by the mouse will not directly correspond to the motion of the robot (and a transformation will be necessary).

> I have only translation fields and no orientation fields

A single mouse is not sufficient to estimate the pose of the robot.
You will need additional sensors (e.g. multiple mouse sensors or an IMU to estimate the orientation).

The dead reckoning produced by the Create 3 is computed by fusing IMU and mouse readings.

**Reply by Rak-r** [reply]
Hi, @alsora , I have been setting up the system now with robot_localization package. I managed to get good values for x and y. One thing I still dont fully understand, that my sensor is mounted at an offset from `base_link` with `x = 0.0, y = 0.4, z = 0.150` and in the `odometry publisher node`, I fill in the odometry message with `frame_id: odom` and `child_frame_id: base_link`. I added the `yaw measurement from my steering angle which might not be the right way but for that I am telling to EKF that not fully trust the `yaw` and give some corrected estimate and then I pass this message to EKF for filtered estimate.

As above discussion, isn't the `robot_state_publisher` publishes the static transform between `optical_sensor_frame` as child and `base_link` as parent, right? So I see the absolute distances in x fairly good as measured them manually as well. 

But for the `yaw` it deviates as the vehicle moves even for straight. I tuned process noise covariance to decrease this error. 

Now I wanted to ask is this setup for `optical_flow_sensor` is correct or do I have to do something more with the transforms? If Yes, how? I am getting a bit confused here. 

This is my complete issue link for more explanation with params and messages output: https://robotics.stackexchange.com/questions/110374/issues-while-setting-up-ekf-for-ackermann-steered-robot
